cpp/mrc/src/internal/service.cpp

```diff
@@ -17,131 +17,295 @@
-            m_state = ServiceState::Stopping;
-        }
+        throw exceptions::MrcRuntimeError(MRC_CONCAT_STR(this->debug_prefix() << " Service must be started before "
+                                                                                 "stopping"));
     }
-    if (execute)
+
+    // Ensure we are at least in the stopping state. If so, execute the stop call
+    if (this->ensure_state(ServiceState::Stopping))
     {
-        do_service_stop();
+        lock.unlock();
+
+        this->do_service_stop();
     }
 }
 
 void Service::service_kill()
 {
-    bool execute = false;
+    std::unique_lock<decltype(m_mutex)> lock(m_mutex);
+
+    if (this->is_service_startable())
     {
-        std::lock_guard<decltype(m_mutex)> lock(m_mutex);
-        if (m_state < ServiceState::Killing)
-        {
-            execute = (m_state < ServiceState::Killing);
-            m_state = ServiceState::Killing;
-        }
+        throw exceptions::MrcRuntimeError(MRC_CONCAT_STR(this->debug_prefix() << " Service must be started before "
+                                                                                 "killing"));
     }
-    if (execute)
+
+    // Ensure we are at least in the stopping state. If so, execute the stop call
+    if (this->ensure_state(ServiceState::Killing))
     {
-        do_service_kill();
+        lock.unlock();
+
+        this->do_service_kill();
     }
 }
 
 void Service::service_await_join()
 {
-    bool execute = false;
     {
-        std::lock_guard<decltype(m_mutex)> lock(m_mutex);
-        if (m_state < ServiceState::Completed)
+        std::unique_lock<decltype(m_mutex)> lock(m_mutex);
+
+        if (this->is_service_startable())
         {
-            execute = (m_state < ServiceState::Completed);
-            m_state = ServiceState::Awaiting;
+            throw exceptions::MrcRuntimeError(MRC_CONCAT_STR(this->debug_prefix() << " Service must be started before "
+                                                                                     "awaiting join"));
         }
-    }
-    if (execute)
-    {
-        do_service_await_join();
-        forward_state(ServiceState::Completed);
-    }
-}
 
-const ServiceState& Service::state() const
-{
-    std::lock_guard<decltype(m_mutex)> lock(m_mutex);
-    return m_state;
-}
+        // Check if this is our first call to service_await_join
+        if (!m_service_await_join_called)
+        {
+            // Prevent reentry
+            m_service_await_join_called = true;
 
-bool Service::is_service_startable() const
-{
-    std::lock_guard<decltype(m_mutex)> lock(m_mutex);
-    return (m_state == ServiceState::Initialized);
+            // We now create a promise and a future to track the completion of the service
+            Promise<void> completed_promise;
+
+            m_completed_future = completed_promise.get_future();
+
+            // Unlock the mutex before calling await join to avoid a deadlock
+            lock.unlock();
```

Comment:
Suggestion: call `lock.unlock()` earlier, since other threads can't enter this block.
```suggestion
            // Unlock the mutex before calling await join to avoid a deadlock
            lock.unlock();

            // We now create a promise and a future to track the completion of the service
            Promise<void> completed_promise;

            m_completed_future = completed_promise.get_future();
```

Reply:
Cant reorder as mentioned in other comment